protocol/app/upgrades/v5.0.0/upgrade.go

```diff
@@ -98,11 +99,41 @@ func blockRateLimitConfigUpdate(
 	)
 }
 
+func negativeTncSubaccountSeenAtBlockUpgrade(
+	ctx sdk.Context,
+	perpetualsKeeper perptypes.PerpetualsKeeper,
+	subaccountsKeeper satypes.SubaccountsKeeper,
+) {
+	// Get block height stored by v4.x.x.
+	blockHeight, exists := subaccountsKeeper.LegacyGetNegativeTncSubaccountSeenAtBlock(ctx)
+	// If no block height was stored in the legacy store, no migration needed.
+	if !exists {
+		return
+	}
+
+	// If there are no perpetuals, then no new state needs to be stored, as there can be no
+	// negative tnc subaccounts w/o perpetuals.
+	perpetuals := perpetualsKeeper.GetAllPerpetuals(ctx)
+	if len(perpetuals) == 0 {
+		return
+	}
+
+	// Migrate the value from the legacy store to the new store.
+	if err := subaccountsKeeper.SetNegativeTncSubaccountSeenAtBlock(
+		ctx,
+		perpetuals[0].Params.Id, // must be a cross-margined perpetual due to `perpetualsUpgrade`.
+		blockHeight,
+	); err != nil {
+		panic(fmt.Sprintf("failed to set negative tnc subaccount seen at block with value %d: %s", blockHeight, err))
+	}
```

Comment:
can we add some log statements to log out the pre and post upgrade data in the event of a success?